ssh: Use paramiko instead of pxssh

douglas-raillard-arm opened the pull request:
Here is an crude port of SshConnection to paramiko, so pxssh can rest in peace.

Quote from a pexpect maintainer in 2014 on https://github.com/DFE/MONK/issues/108:
> As a maintainer of pexpect, I highly recommend using paramiko instead of pexpect for ssh :) pexpect's ssh support is from long ago where no native-ssh option existed, it only remains for compatibility.
With paramiko, remote programs will not detect you as a terminal and will not likely write terminal sequences.

Note: it might be a better idea to just introduce a new class rather than ruthlessly modifying SshConnection, since TelnetConnection relies on pexpect and the recent addition of `options` parameter will not allow easily moving away from an actual ssh/scp command. (paramiko seems to have an OpenSSH config emulation, maybe that would allow that to work ...)

Note2: Some APIs that were not properly abstracted are also somewhat broken, like `background()` method that leaked a `Popen` object, rather than just giving a handle on stdin/stdout/stderr like paramiko does.

All in all, it took 2h to go from scratch to something that seem to work well enough for LISA, which is less than 10% of the time spent chasing TTY size issues (worked around but not really fixed AFAIR) and control sequences issues (https://github.com/ARM-software/devlib/issues/442). It also reduces by 2 the mount of code in `SshConnection`, mostly removing cruft related to interactive terminal management:
```
 2 files changed, 79 insertions(+), 159 deletions(-)
```

douglas-raillard-arm commented:
A quick timeit benchmark on `target.execute('hello world')` on Juno R0 with SSH:
* with pxssh (current devlib): 161 ms ± 456 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
* with paramiko: 36.4 ms ± 4.22 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

That's 4.4 times faster with paramiko. This difference in speed allows a LISA test to go from `2min 30s` to `1min 40s` execution time.

setrofim commented:
Looks good. Thanks for doing this! 

Given the nature of the change, this will need to be thoroughly tested across a range of devices and applications before we merge.

douglas-raillard-arm commented:
Of course, I'm currently testing the following setups:
* debian + arm 64 + password auth
* debian + arm + password auth
* buildroot (OpenSSH_7.8p1) + arm 64 + password auth
* buildroot (OpenSSH_7.8p1) + arm + password auth
* buildroot (OpenSSH_7.8p1) + arm 64 + key auth
* android (adb), which should not be impacted

I expect breakage on those things:
* usage of return value of `target.background()`: It's now a tuple stdin/stdout/stderr but currently is most of the time popen.Popen. After looking around the returned object already differed between `Gem5Connection` and `SshConnection`.
* Users of the newly introduced `options` `__init__` parameter.
* Users expecting an interactive shell (unlikely but who knows ...). Paramiko has a `exec_command(..., get_pty=False)` parameters that might allow that. I would advise against doing that by default, since that's very unlikely to be needed and makes the user's life miserable in most cases. 

TODO:
* `target.background()` needs to allow the user to track the return code of the command, which can be done with the `stdout` object of paramiko but would need to be abstracted away.
* see what can be done for `options`. Paramiko ships an OpenSSH config file parser, but it's up to the user to make use of the key/values. This means we need to know what people will use that for and translate it to parameters to `SSHClient.connect()`. If it's not too latwho knows ...). Paramiko has a `exec_command(..., get_pty=False)` parameters that might allow that. I would advise against doing that by default, since that's very unlikely to be needed and makes the user's life miserable in most cases. 
e, it might be better to remove `options` and replace it with specific supported options that can be implemented for both pxssh and paramiko.
* Check that the exceptions I'm catching are the relevant ones. Paramiko has its own classes, but can also raise exceptions from the std `socket` module like `socket.timeout`.

Opened questions:
The current PR will raise an exception if `sudo` needs to be used when the user only authenticated with a keyfile, since we end up not having any password to feed to sudo. I'm not sure what's the current expected behavior in this case.

setrofim commented:
> The current PR will raise an exception if sudo needs to be used when the user only authenticated with a keyfile, since we end up not having any password to feed to sudo. I'm not sure what's the current expected behavior in this case.

That is probably fine.

> What was process_backspaces() designed for ? If it was only to cope with control sequences, we can safely remove it as I did, since we got a non-interactive shell. If it was there for other reasons, we might want to keep it.

Yeah, it's to deal with TTY weirdness that seemed to leave `^H` characters in the stream on some devices. I'm happy to get rid of it on the assumption that `paramiko` won't present the same issues, and re-introduce this later if it becomes necessary.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -201,185 +192,119 @@ def __init__(self,
-        source = '{}@{}:{}'.format(self.username, self.host, source)
-        return self._scp(source, dest, timeout)
+        sftp = self._get_sftp(timeout)
+        sftp.get(source, dest)
```
When the destination already exists when pushing a folder, should it be wiped or should the new content overlayed on the existing folder ?

marcbonnici replied:
I think overlaying would make sense, this would at least preserve the existing behaviour of scp.

douglas-raillard-arm replied:
Perfect, that's also easier to implement :)

douglas-raillard-arm commented:
Updated with:
* fixed push/pull to work with folders (currently, pushing over an existing folder will write into it rather than wiping it first).
* Fixed locking by using a `threading.RLock` rather than a `threading.Lock`, which allows the lock to be reentrant so we can call `execute()` from anywhere, including to compute properties potentially used by `execute()` itself.

mpekatsoula commented:
> Regarding the SSH options features @mpekatsoula would you be able to provide us with some of your use cases to see how they could be mapped onto this?

@marcbonnici 
We use these options when connecting to a device:
-o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null

setrofim commented:
> -o StrictHostKeyChecking=no

This can be achieved by not loading the host keys (currently done here: https://github.com/ARM-software/devlib/pull/450/files#diff-d0a4376874b01b42b478cb7119010811R206)

> -o UserKnownHostsFile=/dev/null

This handled by the `AutoAddPolicy` set here: https://github.com/ARM-software/devlib/pull/450/files#diff-d0a4376874b01b42b478cb7119010811R207

douglas-raillard-arm commented:
Updating from the offline chat with @setrofim and @marcbonnici about `background()`:

TODO:
1) Implement BackgroundCommand() to abstract over popen.Popen() and whatever paramiko gives to control background commands. It should:
• give a file-like stdout object (containing merged stdout and stderr)
• give a file-like stdin object
• give  a way to send arbitrary POSIX signal
• give a cancel() method with the usual behavior (sigint, then sigkill after a short timeout)
• give a way to wait for completion and return exite code

1.1) Use `BackgroundCommand.cancel()` to implement `SshConnection.cancel_running_command()`

2) Check that the exceptions I'm catching are the relevant ones. Paramiko has its own classes, but can also raise exceptions from the std socket module like socket.timeout.

3) Add an SshConnection parameter equivalent to:
```
    -o UserKnownHostsFile=/dev/null
```
From @setrofim : This handled by the AutoAddPolicy set here: https://github.com/ARM-software/devlib/pull/450/files#diff-d0a4376874b01b42b478cb7119010811R207

4) (optional) see if we can remove the connection locking when executing `background()` if that's reasonably straightforward to do.

This leaves us with the following user-visible change:
* `options` either removed and replaced with ad-hoc parameters. That can be avoided by only using `UserKnownHostsFile` key and raising exception when any other option is found, but this means we will have to exactly match what `ssh` client from OpenSSH gives, otherwise users might (rightfully) complain. If paramiko cannot reproduce the exact behavior, we might run into issues.

* Command will now be ran without a controlling terminal. Do we want to try to support `get_pty=True` ?

douglas-raillard-arm commented:
`background()` question: Paramiko does not seem to allow redirecting stdout/stderr. This means we cannot currently honour these parameters of `background()` unless we spawn a thread that will read the outputs and re-write them into what the user provides. Should we go the side-thread way or should we just drop `stdout` and `stderr` parameters from `background()` ?

That's a bit of a tricky question since that means that the user will *have* to read stdout and stderr, otherwise the command might get blocked until the user does so, leading to a timeout triggering.

setrofim commented:
Getting rid of them is not an option. Can Paramiko be monkey-patched to allow redirection? If not, then go with the thread. From the user's perspective, the behaviour of `background()` should not change from the existing implementation.

setrofim commented:
OK, having had a brief look through Paramiko, monkey-patching isn't going to work, and given its implementation, `subprocess.PIPE`s don't really make much sense. However the blocking on read is potentially a problem, so having a thread drain the underlying channel into a local buffer and then presenting that as a stream might be the way to go.

douglas-raillard-arm commented:
After experimenting a bit, it appears that with the current default `stdout=subprocess.PIPE` already forces the user to consume the output. If that's not done, the command will block if it outputs an implementation-defined amount of data, so all users must consume the output. Since all the APIs I've seen so far force consumption of stdout/stderr, I would say it's reasonable to replicate that in devlib. 

That would mean we only need to spawn a thread if someone gives one of these:
```
* DEVNULL (== read and discard)
* an existing file descriptor (a positive integer)
* an existing file object
* None (== sys.stdout or sys.stderr)
```
(list taken from https://docs.python.org/3/library/subprocess.html#subprocess.Popen)

douglas-raillard-arm commented:
Updated with `BackgroundCommand()` abstract base class, mirroring a subset of `subprocess.Popen` API. It's implemented on top of both `subprocess.Popen` and `paramiko` objects.

remaining TODOs:
0) Implement custom stream redirection in `SshConnection.background()` with a companion thread to copy data from paramiko into another file-like object.
1) Use `BackgroundCommand.cancel()` to implement `SshConnection.cancel_running_command()`
2) Check that the exceptions I'm catching are the relevant ones. Paramiko has its own classes, but can also raise exceptions from the std socket module like socket.timeout.
3) Add option to implement  `-o UserKnownHostsFile=/dev/null`
4) (optional) see if we can remove the connection locking when executing background() if that's reasonably straightforward to do.

douglas-raillard-arm commented:
@marcbonnici @setrofim The PR should now be finished feature-wise and somewhat ready for testing. It will need some polishing for review (I'll try to break down the big commit in smaller ones), but the code itself can now be expected to work.

The main changes are:
1. `BackgroundCommand()` offering a `subprocess.Popen`-like interface, and is expected to behave as such. Differences between an actual Popen should be looked into if they are fixable. Main differences are:
* Actions done on the `BackgroundCommand()` are "transparent" so that they implact the remote command, not whatever local proxy command in use (ex `adb shell`).
* sending a signal sends it to the whole process group rather than just one process. That makes the code work regardless of the nesting of `sh`/`sudo` calls, non-blocking `&` execution etc.
* Closing `stdout` on an SSH remote command will not send SIGPIPE to the remote command when it tries to read it until stderr is closed as well. (if it was redirected to `subprocess.DEVNULL` it's considered closed). This is because of a limitation of SSH which uses one channel multiplexing stdin/stdout/stderr, and only closing that channel ends up in closing the writer end of the pipe of the remote process.

2. `options` is gone, replaced by `strict_host_check=True`. When `True`, any unknown server will be rejected, when `False` unknown servers will be added to known_hosts. 

3. Any number of concurrent `target.background()` commands can be started at the same time. `target.cancel_running_command()` will call `BackgroundCommand.cancel()` on all the ones that are still alive, to send them SIGTERM then SIGKILL.

4. Commands ran with SSH don't have a controlling terminal, so interactive programs won't work. `sudo -S` is used to read the password from stdin instead. Output lines will not wrap anymore and most commands will stop sending colored output.

Performance-wise, an `target.execute('echo hello world')` now takes ~30ms roundtrip compared to ~160ms with pxssh.

douglas-raillard-arm commented:
@marcbonnici @setrofim I've split the PR into more contained commits. The last one is pretty big but there isn't really any alternative since it's almost all new code.

It should now be ready for review proper and testing so I removed the [RFC] tag on the PR name.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,422 @@ def name(self):
+
+        if use_sudo:
+            if not self.password:
+                raise TargetStableError('Attempt to use sudo but no password was specified')
```
No password is prompted since that's not an interactive shell. This means we use `sudo -S` to read the password from stdin. We could make check for `None` though:
```
if self.password is None:
     raise ....
```
Which will allow passing an empty string as password.

marcbonnici replied:
That seems like it might be a better option. Another thought out of curiosity, do you know what state occurs if we just don't provide the password after invoking sudo (if no password is provided) and is that state easy to detect?

douglas-raillard-arm replied:
Good question, now that I think about it, passwordless sudo might just execute the command without even looking at stdin, in which case we must not supply one (otherwise the called command will receive a newline it never asked for). Do you have  a passwordless system where you can check the behavior of `sudo -S` ?

marcbonnici replied:
Just checked on my chrombook and that is the case, it immediately just executes the command.

douglas-raillard-arm replied:
So now the situation is:
* password=None means we don't provide a password. If as_root is used, it will raise an exception
* password="" means sudo is configured to run passwordless. If sudo actually needs a password, it will stall waiting on stdin.

The only way to avoid the stall in case of misconfiguration would be to send "\n" in all cases. Unfortunately, that means sending a newline to a command that did not ask for it when everything is well configured. All that can be avoided by a check at connection time to see if sudo is passwordless or not. Should we go that way ?

marcbonnici replied:
If we can check this once at connection time that seems like a good plan.
Although if we do this check, does it still make sense to make the distinction between providing an empty password vs no password?

douglas-raillard-arm replied:
I think it's a good idea to allow users to disable the use of sudo if they want to. That could be used to prevent devlib from poking into important things without knowing about it.

marcbonnici replied:
Ah ok I hadn't thought about using that as a feature.

In which case just to clarify the above:
> password=None means we don't provide a password. If as_root is used, it will raise an exception

Will this still allow for the usecase where we're connected as root (using a private key, no password) and calling `execute` with `as_root=True` since then `sudo` wouldn't be required?

douglas-raillard-arm replied:
good point, `password` is not a good indicator of what the user wants. I guess we should just check if it's passwordless sudo or not and just check for `if self.password:` as you suggested.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,422 @@ def name(self):
+            bg_cmds = set(self._current_bg_cmds)
+            for bg_cmd in bg_cmds:
+                bg_cmd.close()
+            self.client.close()
```
Good point, I used `__del__` and I guess that's one more occurence of someone being bitten by it. The exception should be harmless though as it's swallowed by the interpreter when coming from `__del__`. Maybe we should hide `AttributeError` explicitly to avoid ugly messages being printed.

marcbonnici replied:
I'm not sure sure about just hiding the error as this may lead to confusion in the future. Would it be worth declaring the attribute explicitly as None before initialising the connection and checking if the client exists before attempting to close it?

douglas-raillard-arm replied:
If we want to avoid calling `close()` until `__init__` is finished, we could add a checkpoint at the end of `__init__` with a metaclass quite easily, so `__del__` would now what is safe to do. This would work well with subclassing, and would not be susceptible to mistakes in the future when someone modifies the classes.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,422 @@ def name(self):
+                bg_cmd.close()
+            self.client.close()
+
+    def _execute_with_sudo(self, command, as_root, log, timeout, executor):
```
I was out of inspiration. The name comes from the fact that it's the function that will add `sudo` if needed, do you have an idea/convention to follow ?

marcbonnici replied:
That makes sense, I just found it a bit unclear what the purpose was only by its name next to a method without sudo in its name. How about something like `_execute_command` just to be a but more generalised?

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,422 @@ def name(self):
+        select_timeout = 1
+        output_chunks, exit_code = _read_paramiko_streams(stdout, stderr, select_timeout, callback, [])
+        # Join in one go to avoid O(N^2) concatenation
+        output = b''.join(output_chunks)
```
The streams from the command itself are actually merged by:
```

        # Merge stderr into stdout since we are going without a TTY
        command = '({}) 2>&1'.format(command) 
```
The setup you have seems to have flaky configuration, I'm not sure it's a good idea to get rid of stderr on some arbitrary criteria like return code == 0. Even without considering that, this behavior would be impossible to implement on adb, meaning any code written against an SSH connection relying on that would stop working on some other connection types.

For now, the `execute()` method merges stdout and stderr, if something else is needed `background()` can be used for more manual process control. It does allow separate handling of stdout/stderr.

That said we could add a "discard_stderr=False" option to `execute()`, which should be implementable for all connection types. It won't swallow shell config problems like you have (unless we wrap everything with a shell command that does not read any config and that discards stderr).

marcbonnici replied:
Agreed that my setup is not quite correctly configured however in this case I don't think the std error is coming from the command invocation and therefore is not merged initially, this is coming from the terminal itself, as you can see we're getting separate `output_chunks`.

I'm sure also not sure that disregarding it would make sense or what the behaviour should be. Just wanted to highlight the behaviour I'd noticed so far is unclear, this manifested as a `FileNotFound` error when attempting to use a path discovered from the device (since the path also included this error message).

douglas-raillard-arm replied:
The same would happen if e.g. a command is not found. There is no way to tell the difference between something written on stderr by the shell and something written by the command you care about. The way most APIs like `subprocess` handle that is by making stderr opt-in, which is usually what people care about. In devlib case, `execute()` merges both as things like adb don't support splitting the two.

The only way of "generically" fixing that problem is changing `execute()` semantic to discard stderr by default, which is an API break but may actually improve reliability of most existing code.

marcbonnici replied:
That makes sense and I think you're right, breaking the API or trying to hide the output arbitrarily probably isn't the best idea. I guess if possible the underlying problem with the environment should just be fixed (simple enough for my example case) and if not then it could be dealt with explicitly for the particular device.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,422 @@ def name(self):
+            cls._pull_file(sftp, src, dst)
+        except IOError:
+            # Maybe that was a directory, so retry as such
+            cls._pull_folder(sftp, src, dst)
```
Rewording to make sure I understand correctly:
* _pull_file or _pull_folder will be selected based on what there is to pull
* if a file is pulled in a folder, the src basename will be appended to the dst folder

marcbonnici replied:
- Agreed, the existing functionality seems to identify this correctly.
- That at least is what makes sense to me to mimic existing behaviour of the other connections. For example within WA we frequently just want to pull file X into our output directory rather than explicitly creating the full file paths for both source and destination.

douglas-raillard-arm replied:
Should pushing an src folder into a dst folder do:
1) Push the content of src into dst
2) "move" src into dst, which means creating dst/src.basename

The 1st behavior seems more sensible and make pushing folder idempotent, which I think is very desirable. The 2nd behavior corresponds to the mv command.

marcbonnici replied:
Good question. The first option might seem useful however just looking at the other implementations and it looks like the 2nd option is more consistent with the `AdbConnection` and the existing `SshConnection` behaviour so unsure it would make sense for this version to operate differently?

douglas-raillard-arm replied:
Looks like adb actually behaves like 1):
```
$ adb push . /mnt/sdcard/
push: ./t2.txt -> /mnt/sdcard/t2.txt
push: ./t3.txt -> /mnt/sdcard/t3.txt
push: ./t1.txt -> /mnt/sdcard/t1.txt
3 files pushed. 0 files skipped.
```
https://stackoverflow.com/questions/13145872/android-recursive-copy-with-adb-push

douglas-raillard-arm commented on devlib/utils/misc.py:
```diff
@@ -183,27 +179,25 @@ def callback(pid):
+    else:
+        timeout_expired = None
+
+    if sys.version_info[0] == 3:
```
Probably not, that's copied from the current code. Devlib now only supports 3.5 is that right ?

marcbonnici replied:
Yea that's correct, just thinking might be a good opportunity to remove this since there are other Python3 specific code here now.

douglas-raillard-arm commented on devlib/utils/ssh.py:
```diff
@@ -170,51 +305,473 @@ def name(self):
+                'StrictHostKeyChecking': 'no',
+                'UserKnownHostsFile': '/dev/null',
+            }
+        self.options = options
```
I'm not sure, the goal was to try to reproduce the current behavior since I don't have any telnet target at hand. If you think it's useless I can just remove it

douglas-raillard-arm replied:
made me realize that I introduced a copy of `ssh_get_shell`, so I renamed it into `telnet_get_shell` and specialized it for telnet by removing irrelevant parameters.

setrofim commented on devlib/connection.py:
```diff
@@ -0,0 +1,326 @@
+    def stderr(self):
+        """
+        File-like object connected to the background's command stderr.
+        """
```
Add `pid` as a property as well.

douglas-raillard-arm replied:
Done

setrofim replied:
:+1:

douglas-raillard-arm commented:
Hi @setrofim @marcbonnici, should I squash the WIP patches into the ones they are fixing up to prepare the final version of the branch ?

setrofim commented:
Yes, please. I think we've gotten all the feedback we're going to for this; and I don't think anything more would need to be changed.

douglas-raillard-arm commented:
@setrofim the branch is updated
